### ISSUE TITLE
YOLO V5 for custom training and detection

### ISSUE THREAD

**Issue by bugmany** [ultralytics/yolov5#2983]
## ❔Question
1.train:
python3 train.py --data custom.yaml --cfg yolov5s.yaml --weights weights/yolov5s.pt --epochs 5
train: Scanning 'data/custom/labels/train.cache' images and labels... 6 found, 0 missing, 0 empty, 0 corrupted: 100%|█| 6/6 [0
val: Scanning 'data/custom/labels/train.cache' images and labels... 6 found, 0 missing, 0 empty, 0 corrupted: 100%|█| 6/6 [00:Plotting labels... 
val: Scanning 'data/custom/labels/train.cache' images and labels... 6 found, 0 missing, 0 empty, 0 corrupted: 100%|█| 6/6 [00:

autoanchor: Analyzing anchors... anchors/target = 3.86, Best Possible Recall (BPR) = 1.0000
Image sizes 640 train, 640 test
Using 4 dataloader workers
Logging results to runs/train/exp3
Starting training for 5 epochs...

     Epoch   gpu_mem       box       obj       cls     total    labels  img_size
       0/4        0G    0.1254   0.04242   0.08415     0.252        37       640: 100%|█████████| 1/1 [00:05<00:00,  5.10s/it]
               Class      Images      Labels           P           R      mAP@.5  mAP@.5:.95: 100%|█| 1/1 [00:03<00:00,  3.89s
                 all           6           0           0           0           0           0
Images sizes do not match. This will causes images to be display incorrectly in the UI.

     Epoch   gpu_mem       box       obj       cls     total    labels  img_size
       1/4        0G    0.1246   0.04416   0.08551    0.2543        42       640: 100%|█████████| 1/1 [00:06<00:00,  6.93s/it]
               Class      Images      Labels           P           R      mAP@.5  mAP@.5:.95: 100%|█| 1/1 [00:03<00:00,  3.41s
                 all           6           0           0           0           0           0

     Epoch   gpu_mem       box       obj       cls     total    labels  img_size
       2/4        0G    0.1251   0.05065   0.08903    0.2648        46       640: 100%|█████████| 1/1 [00:05<00:00,  5.33s/it]
               Class      Images      Labels           P           R      mAP@.5  mAP@.5:.95: 100%|█| 1/1 [00:03<00:00,  3.72s
                 all           6           0           0           0           0           0

     Epoch   gpu_mem       box       obj       cls     total    labels  img_size
       3/4        0G   0.07845   0.04337   0.05434    0.1762        36       640: 100%|█████████| 1/1 [00:05<00:00,  5.19s/it]
               Class      Images      Labels           P           R      mAP@.5  mAP@.5:.95: 100%|█| 1/1 [00:03<00:00,  3.28s
                 all           6           0           0           0           0           0

     Epoch   gpu_mem       box       obj       cls     total    labels  img_size
       4/4        0G    0.1235    0.0542   0.08529     0.263        54       640: 100%|█████████| 1/1 [00:04<00:00,  4.73s/it]
               Class      Images      Labels           P           R      mAP@.5  mAP@.5:.95: 100%|█| 1/1 [00:03<00:00,  3.22s
                 all           6           0           0           0           0           0
5 epochs completed in 0.014 hours.

Optimizer stripped from runs/train/exp3/weights/last.pt, 14.5MB
Optimizer stripped from runs/train/exp3/weights/best.pt, 14.5MB
Images sizes do not match. This will causes images to be display incorrectly in the UI.

wandb: Waiting for W&B process to finish, PID 79757
wandb: Program ended successfully.

2.detect:
python3 detect.py --source data/images --weights weights/best.pt --conf 0.25
Fusing layers... 
Model Summary: 224 layers, 7113244 parameters, 0 gradients, 16.5 GFLOPS
image 1/2 /Users/home/dev/yolov5/data/images/bus.jpg: 640x480 Done. (0.182s)
image 2/2 /Users/home/dev/yolov5/data/images/zidane.jpg: 384x640 Done. (0.148s)
Results saved to runs/detect/exp2
Done. (0.400s)

at last, i did not detect anything information with my train models ("best.pt"), can anyone help me?

**Comment by glenn-jocher** [comment]
@bugmany 👋 Hello! Thanks for asking about improving training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥1.5k images per class
- **Instances per class.** ≥10k instances (labeled objects) per class total
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x will produce better results in nearly all cases, but have more parameters and are slower to run. For mobile applications we recommend YOLOv5s/m, for cloud or desktop applications we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

To start training from pretrained weights simply pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
```

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://user-images.githubusercontent.com/26833433/103595982-ab986000-4eb1-11eb-8c57-4726261b0a88.png"></p>

## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by bugmany** [comment]
hello @github-actions and @glenn-jocher, 
thanks for yours reply, 

here is W&B info:
[debug.log](https://github.com/ultralytics/yolov5/files/6431514/debug.log)
[debug-internal.log](https://github.com/ultralytics/yolov5/files/6431515/debug-internal.log)

reproduce this issue:
git path: 
https://github.com/bugmany/yolov5_bm.git
train command:
python3 train.py --data custom.yaml --cfg yolov5s.yaml --weights weights/yolov5s.pt --img 640 --epochs 5
detect command:
python3 detect.py --source data/images --weights weights/best.pt --conf 0.25

thanks a lot.

**Comment by bugmany** [comment]
I 've solved it by increased epochs times.
